docs/versioning.md

```diff
@@ -0,0 +1,488 @@
+should exist at any given time, and it can be identified in S3 requests
+and responses with the version id "null".
+
+With respect to metadata, since the null version is overwritten by subsequent
+null versions, the null version is initially stored in the master key alone,
+as opposed to being stored in the master key and a new version, like how normal
+versions are stored. The metadata engine does not differentiate between null and
+normal versions, so S3 checks if versioning is suspended or has never been
+configured, and sets the `versionId` option to `''` in PUT calls to the metadata
+engine.
+
+The tables below show the keys resulting from PUT calls to metadata if we put an
+object 'foo' twice, when versioning has not been enabled or is suspended.
+
+(1) PUT foo (first put), versionId: `''`
+
+| key | value |
+|-----|:-----:|
+| foo (null) |   A   |
+
+(2) PUT foo (second put), versionId: `''`
+
+| key | value |
+|-----|:-----:|
+| foo (null) |   B   |
+
+However, null versions are preserved when versioning has been enabled or
+re-enabled, and new non-null versions are created. To preserve the null version
+if the null version is the master version, the the S3 API stores the current
+null version as a new key `(3A)` prior to overwriting the master version `(3B)`.
+This implies the null version may not necessarily be the latest or master
+version.
+
+To continue the example immediately above, if we enabled versioning and
+put another object, the calls to metadata and resulting keys would resemble the
+following:
+
+(3A) PUT foo, versioning: `true`
+
+| key | value |
+|-----|:-----:|
+| foo |   B   |
+| foo.v1 (null) |   B   |
+
+(3B) PUT foo, versioning: `true`
+
+| key | value |
+|-----|:-----:|
+| foo |   C   |
+| foo.v2 |   C   |
+| foo.v1 (null) |   B   |
+
+Now, in the case versioning is suspended and the null version is not the latest
+version, the S3 API sends a delete instruction to metadata to delete the
+current null version when creating a new null version.
+
+(4A) DELETE foo, versionId: `v1` (null version ID)
+
+| key | value |
+|-----|:-----:|
+| foo |   C   |
+| foo.v2 |   C   |
+
+Then the master version is overwritten with the new null version:
+
+(4B) PUT foo, versionId: `''`
+
+| key | value |
+|-----|:-----:|
+| foo (null) |   D   |
+| foo.v2 |   C   |
+
+Note that S3 keeps track of the internally-used version ID of the current null
+version in the `nullVersionId` metadata attribute for each new version.
+In addition, to determine whether the null version is the latest version, the S3
+API sets the `isNull` attribute for each version. The `nullVersionId` attribute
+is also used to retrieve the correct version when the version ID "null" is
+specified in certain object-level APIs, described further in the section
+"Null Version Mapping".
+
+#### Specifying Versions in APIs for Putting Versions
+
+Since S3 does not currently allow overwrite of existing version data, Put
```

Comment:
There's some code in the versioningPreprocessing function that deals with scenarios where we would be overwriting existing version data. I think we wrote it in case it would be needed for bucket replication, but at the moment we return 400 InvalidArgument before the request reaches versioningPreprocessing.